role/processor_test.go

```diff
@@ -559,3 +598,708 @@ func Test_rtp_keyFunc(t *testing.T) {
+					UserID:   "domain.tenant.service",
+					Scope:    []string{"admin", "user"},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "verify certificate bound accecss token fail, cnf check fail and no client_id",
+			fields: fields{
+				jwkp: jwk.Provider(func(kid string) interface{} {
+					return LoadRSAPublicKeyFromDisk("./asserts/public.pem")
+				}),
+				enableMTLSCertificateBoundAccessToken: true,
+			},
+			args: args{
+				cert: &x509.Certificate{
+					KeyUsage: x509.KeyUsageDigitalSignature,
+					Subject: pkix.Name{
+						CommonName: "dummy",
+					},
+					NotBefore: time.Now(),
+					NotAfter:  time.Unix(9999999999, 0),
+				},
+				claims: &ZTSAccessTokenClaim{
+					BaseClaim: BaseClaim{
+						StandardClaims: jwt.StandardClaims{
+							Subject:   "domain.tenant.service",
+							IssuedAt:  1585122381,
+							ExpiresAt: 9999999999,
+							Issuer:    "https://zts.athenz.io",
+							Audience:  "domain.provider",
+						},
+					},
+					AuthTime: 1585122381,
+					Version:  1,
+					UserID:   "domain.tenant.service",
+					Scope:    []string{"admin", "user"},
+					Confirm:  map[string]string{"x5t#S256": "2jt82f2uM8jE2LMcb4erhhTc-uy1yB1iEyp5MnI5uF4"},
+				},
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := &rtp{
+				pkp:                                   tt.fields.pkp,
+				jwkp:                                  tt.fields.jwkp,
+				enableMTLSCertificateBoundAccessToken: tt.fields.enableMTLSCertificateBoundAccessToken,
+				clientCertificateGoBackSeconds:        tt.fields.clientCertificateGoBackSeconds,
+				clientCertificateOffsetSeconds:        tt.fields.clientCertificateOffsetSeconds,
+			}
+			if err := r.validateCertificateBoundAccessToken(tt.args.cert, tt.args.claims); (err != nil) != tt.wantErr {
+				t.Errorf("rtp.validateCertificateBoundAccessToken() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func Test_rtp_validateCertPrincipal(t *testing.T) {
+	type fields struct {
+		pkp                                   pubkey.Provider
+		jwkp                                  jwk.Provider
+		enableMTLSCertificateBoundAccessToken bool
+		clientCertificateGoBackSeconds        int64
+		clientCertificateOffsetSeconds        int64
+	}
+	type args struct {
+		cert   *x509.Certificate
+		claims *ZTSAccessTokenClaim
+	}
+	LoadRSAPublicKeyFromDisk := func(location string) *rsa.PublicKey {
+		keyData, e := ioutil.ReadFile(location)
+		if e != nil {
+			panic(e.Error())
+		}
+		key, e := jwt.ParseRSAPublicKeyFromPEM(keyData)
+		if e != nil {
+			panic(e.Error())
+		}
+		return key
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "verify cert principal success",
+			fields: fields{
+				jwkp: jwk.Provider(func(kid string) interface{} {
+					return LoadRSAPublicKeyFromDisk("./asserts/public.pem")
+				}),
+				enableMTLSCertificateBoundAccessToken: true,
```

Comment:
```suggestion
```
not used?

Reply:
43f2cf4acb894e171f7e7efd61d3a7a4368a7cc3